CMakeLists.txt

```diff
@@ -115,6 +115,7 @@ include(CheckCXXCompilerFlag)
 set(COMMON_CXX_FLAGS
   -pipe
   -Wall
+  -march=native
```

Comment:
This should be for release flags only, not the common ones.

Also, we may want to use `-msse4.1` instead of `-march=native` here:  if you use `-march=native` then the binary compiled for one machine may not run on another machine.

Reply:
Yes, it should be for release flags.
I use the flag `march=native` instead of specific flags to let compiler uses the most efficient one. `simd.h` will choose the appropriated files, e.g. `simd_128.h` for now, or `simd_256.h` (later) for AVX2 etc.

Reply:
I think we will need a proper framework to select the most-efficient one (and to have a fallback for CPU with no SIMD), it won't be solved by `march=native` alone.
It's better to select the most optimized version dynamically (using CPU dispatch once, at the lib initialization) rather than using compile-time settings (that's how most of Intel libs works).
We don't have such "framework" for now, but we should introduce one at one point.

IMHO, if we want to compile with `march=native` it should be done by enabling an option.
Generating non-portable binary by default hurt the reproducibility of our benchmark/tests/etc.
It can be OK for local dev/test, but no good for CI & cie.

Reply:
I like that framework :)
So maybe we should have an option to generating binaries with SIMD instruction?

Reply:
Yes, we could add an option to CMake (USE_SIMD, or USE_HW_ACCELERATION) that would be True by default (but could be disabled on-demand, if people compile on old/unsupported CPU).
When this option is set, we would have a preprocessor symbol defined, and then we can use #ifdef in our code to include or not SIMD code according to this.

Latter, we can add a "framework" for runtime selection, but anyway SIMD should be optional so we will need a compile-time flag for it (can be done in another PR though, this PR is mainly about introducing your SIMD code).